Show mention indicator

Sohaib-Ahmed21 opened the pull request:
<!-- Describe your pull request here.-->
This PR allows showing the special @ indicator for unread @-mentions in DMs in Recent conversations and Inbox in left sidebar which was only showed before in Mentions in left sidebar.
I have added these functionalities:
1. Takes related unread message ids (from inbox and recent_view  javascript files) in specific context from zulip/web/src/unread.ts.
2. Takes unread messages ids and checked if the dms have mention from zulip/web/src/util.ts which checks presence of mention in each message separately from message_has_mention function from zulip/web/src/unread.ts .
3. Then rendered mention indicator on front-end based on presence of mention in dm.

Fixes: #28849 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
[Showing_mention_indicators.webm](https://github.com/zulip/zulip/assets/113179991/8f5244f1-7c54-4b53-8218-700b93da7081)

Mention indicators in light mode:

![Light-mode-inbox-view](https://github.com/zulip/zulip/assets/113179991/f0eedd4c-690e-4851-855b-ec1cc519331f)

![Light-mode-Recent-conversations](https://github.com/zulip/zulip/assets/113179991/33155cb3-770b-46bc-ae4c-24ec7489cd4e)

Mention indicators in dark mode:

![Dark-mode-Inbox-view](https://github.com/zulip/zulip/assets/113179991/92468693-dec8-4ed7-8b84-fa9e3655cb61)


![Dark-mode-Recent-conversations](https://github.com/zulip/zulip/assets/113179991/d8dd40c5-65c0-41c7-ba40-fe5aab534d65)




<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Highlights technical choices and bugs encountered.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.

</details>

alya commented:
Thanks! To get feedback on your PR, you will need to post screenshots (in light and dark theme), not just a video, as indicated in the PR template. Your PR will also need to pass tests.

alya commented:
Please clean up your commit history and post again to request a review. See [here](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html) for guidelines.

Sohaib-Ahmed21 commented:
I am working on cleaning commit history.

Sohaib-Ahmed21 commented:
I have cleaned up the commit history. Kindly review.

Sohaib-Ahmed21 commented:
@alya kindly review this PR.

Sohaib-Ahmed21 commented:
@alya kindly review this PR.

alya commented:
Thanks!  The alignment doesn't look right, e.g., definitely too low in Inbox. Please update the PR to match @-mentions y  streams, and then update the screenshots to include both, so that one can see that they look the same.

Sohaib-Ahmed21 commented:
@alya now I have added the functionality of showing @-mention-indicator for DMs in left sidebar which was not added before and also now updated the screenshots showing stream @-mention-indicators along side newly implemented mention-indicators to see the consistency of newly implemented ones. Kindly review.

alya commented:
I don't see left sidebar @-mention indicators or stream @-mention indicators in the screenshots.

Sohaib-Ahmed21 commented:
@alya Extremely sorry I forgot to save the changes in pull request. Now I have updated screenshots.

alya commented:
Thanks! @amanagr could you please review this one? I haven't tested.

amanagr commented on web/src/inbox_ui.js:
```diff
@@ -174,6 +174,9 @@ function format_dm(user_ids_string, unread_count, latest_msg_id) {
         is_hidden: filter_should_hide_row({dm_key: user_ids_string}),
         is_collapsed: collapsed_containers.has("inbox-dm-header"),
         latest_msg_id,
+        mention_in_unread: unread.dms_has_mention(
+            unread.get_msg_ids_for_user_ids_string(user_ids_string),
+        ),
```
Let's use a name that specifies that this is just for dms like `mention_in_unread_dms`.

amanagr replied:
This needs to be done for all your changes, it make the behaviour buggy when there are mention in topics.

Sohaib-Ahmed21 replied:
yeah that makes sense, thanks. Implementing the changes

amanagr commented on web/src/inbox_ui.js:
```diff
@@ -1013,6 +1018,9 @@ export function update() {
         $inbox_dm_header.removeClass("hidden_by_filters");
+        const mention_in_unread = unread.dms_has_mention(unread.get_msg_ids_for_dms());
+        $inbox_dm_header.find(".unread_mention_info").text(mention_in_unread ? "@" : "");
+        $inbox_dm_header.find(".unread_count").text(unread_dms_count);
         $inbox_dm_header.find(".unread_count").text(unread_dms_count);
```
Repeated.

Sohaib-Ahmed21 replied:
Oh, I messed up, correcting it.

amanagr commented on web/src/unread.ts:
```diff
@@ -964,6 +964,17 @@ export function get_msg_ids_for_topic(stream_id: number, topic_name: string): nu
+
+    for (const msg_id of idsToCheck) {
+        if (message_has_mention(msg_id)) {
+            return true; // Return true if any msg_id returns true from message_has_mention
+        }
+    }
+    return false; // Return false if none of the msg_ids return true from message_has_mention
```
These comments are not helpful so we can remove them.

Sohaib-Ahmed21 commented:
@amanagr I have updated the commits as per the review. Kindly review.

amanagr commented:
LGTM! Thanks

Sohaib-Ahmed21 commented:
@alya kindly review this.

alya commented:
The "integration review" label means that reviewing this PR is on @timabbott 's todo list.

Sohaib-Ahmed21 commented:
> The "integration review" label means that reviewing this PR is on @timabbott 's todo list.

Oh! sorry for mistaken mention. @timabbott kindly review it.

timabbott commented on web/src/unread.ts:
```diff
@@ -964,6 +964,17 @@ export function get_msg_ids_for_topic(stream_id: number, topic_name: string): nu
 }
 
+export function dms_has_mention(msg_ids: number | number[]): boolean {
+    const idsToCheck = Array.isArray(msg_ids) ? msg_ids : [msg_ids];
```
This variable name doesn't follow our snake case coding style guidelines.

timabbott replied:
Also I think this function signature should probably just accept an array of `message_ids`

timabbott commented on web/templates/inbox_view/inbox_list.hbs:
```diff
@@ -9,6 +9,9 @@
                 </div>
+                <span class="unread_mention_info tippy-zulip-tooltip"
+                  {{#unless mention_in_unread_dms}}hidden{{/unless}}
+                  data-tippy-content="{{t 'You have mentions'}}">@</span>
```
This doesn't match the whitespace for other translation tags.

timabbott commented on web/templates/pm_list_item.hbs:
```diff
@@ -13,8 +13,17 @@
+                </span>
+            {{else if is_followed}}
+                <i class="zulip-icon zulip-icon-follow" aria-hidden="true"></i>
+            {{/if}}
```
I think you don't need this, since you can't follow a DM conversation.

timabbott commented on web/styles/left_sidebar.css:
```diff
@@ -974,6 +974,26 @@ li.top_left_scheduled_messages {
+        /* Height is set here by the flexbox; this
+           decouples .unread_count from the app-wide
+           definition. */
+        height: 15px;
```
Why is this 15px, rather than the value of the very similar rules for `.stream-markers-and-controls,` and friends? 

Also, is there a reason this is separate from that rule, or at least not next to it?

timabbott commented on web/tests/pm_list_data.test.js:
```diff
@@ -172,6 +178,7 @@ test("get_conversations bot", ({override}) => {
             status_emoji_info: undefined,
             is_group: true,
             is_bot: false,
+            dms_contain_unread_mention: false,
```
Can you extend the test setup such that one of these cases has this value `true`?

timabbott commented on web/src/unread.ts:
```diff
@@ -964,6 +964,17 @@ export function get_msg_ids_for_topic(stream_id: number, topic_name: string): nu
+    }
+    return false;
+}
+
```
Architecturally, I'm not sure about the design here; it seems like it might be cleaner to have this code path compute which DM conversations contain mentions when processing everything in `unread.ts`, rather than having these follow-up calls, each of which do a nontrivial loop:

```
    get_counts(include_latest_msg_id = false): DirectMessageCountInfo {                                 
        const pm_dict = new Map(); // Hash by user_ids_string -> count Optional[, max_id]               
        let total_count = 0;                                                                            
        for (const [user_ids_string, id_set] of this.bucketer) {                                        
            const count = id_set.size;                                                                  
            if (include_latest_msg_id) {                                                                
                const latest_msg_id = Math.max(...id_set);                                              
                pm_dict.set(user_ids_string, {                                                          
                    count,                                                                              
                    latest_msg_id,                                                                      
                });                                                                                     
            } else {                                                                                    
                pm_dict.set(user_ids_string, count);                                                    
            }                                                                                           
            total_count += count;                                                                       
        }                                                                                               
        return {                                                                                        
            total_count,                                                                                
            pm_dict,                                                                                    
        };                                                                                              
    }                                                                                                   
```

Sohaib-Ahmed21 commented:
Thanks @timabbott for the detailed review. I will address the changes and will respond shortly. Actually work on some other projects is also going on. So addressing reviews here might take some time, 2 days probably.

Sohaib-Ahmed21 commented:
Currently working on addressing reviews.

Sohaib-Ahmed21 commented:
My VM ran into an issue and I am looking forward to addressing these reviews as soon as VM gets working properly.

alya commented:
@Sohaib-Ahmed21 Do you still plan to complete the work here, or should someone else take over finishing this PR?

Sohaib-Ahmed21 commented:
> @Sohaib-Ahmed21 Do you still plan to complete the work here, or should someone else take over finishing this PR?

I have told @timabbott on Zulip dm about my academic activities and that completion label would be good on my opened PRs based on my current workload.

I tried to manage working on these PRs but couldn't spare time to work due to multiple projects overload.

timabbott commented:
Closing in favor of #30204 for this issue.
